spec-docs/function-types.md

```diff
@@ -0,0 +1,304 @@
+(throwing exceptions), and which knows its arity.
+Kotlin lambdas are translated to subclasses of this abstract class, passing the correct arity to the super constructor.
+* Provide a way to get arity of an arbitrary `Function` object (pretty straightforward).
+* Hack `is/as Function5` on any numbered function in codegen (and probably `KClass.cast()` in reflection) to check against `Function` and its arity.
+
+## Extension functions
+
+Extension function type `T.(P) -> R` is now just a shorthand for `@kotlin.extension Function2<T, P, R>`.
+`kotlin.extension` is a **type annotation** defined in built-ins.
+So effectively functions and extension functions now have the same type,
+which means that everything which takes a function will work with an extension function and vice versa.
+
+To prevent unpleasant ambiguities, we introduce additional restrictions:
+* A value of an extension function type cannot be **called** as a function, and a value of a non-extension
+function type cannot be called as an extension. This requires an additional diagnostic which is only fired
+when a call is resolved to the `invoke` with the wrong extension-ness.
+* If an extension function **literal** argument has some shape (its parameters are written out explicitly
+in the code and it's evident that there is or there isn't a receiver parameter), this shape must exactly match
+the extension-ness of the corresponding parameter. You can't pass an extension function **literal**
+where a function is expected and vice versa. The same holds for function expressions.
+If you really want to do that, change the shape or use the `as` operator.
+
+So basically you can now safely coerce values between function and extension function types,
+but still should invoke them in the format which you specified in their type (with or without `@extension`).
+
+With this we'll get rid of classes `ExtensionFunction0`, `ExtensionFunction1`, ...
+and the rest of this article will deal only with usual functions.
+
+## Function0, Function1, ... types
+
+The arity of the functional interface that the type checker can create in theory **is not limited** to any number,
+but in practice should be limited to 255 on JVM.
+
+These interfaces are named `kotlin.Function0<R>`, `kotlin.Function1<P0, R>`, ..., `kotlin.Function42<P0, P1, ..., P41, R>`, ...
+They are *fictitious*, which means they have no sources and no runtime representation.
+Type checker creates the corresponding descriptors on demand, IDE creates corresponding source files on demand as well.
+Each of them inherits from `kotlin.Function` (described below) and contains only two functions,
+both of which should be synthetically produced by the compiler:
+* (declaration) `invoke` with no receiver, with the corresponding number of parameters and return type.
+* (synthesized) `invoke` with first type parameter as the extension receiver type, and the rest as parameters and return type.
```

Comment:
Then I do not understand the shape of the `invoke` method descibed here. Could you give an example?